### PR TITLE
[`ruff`] Implement `quadratic-list-summation` rule (`RUF017`)

### LINKED ISSUE #5073: Warn against quadratic list summation
I can't find a linter that warns me about `sum(list_of_lists, [])`. Basically any time the `default` arg of `sum` is a list literal, I think something slow could happen. Maybe ruff could warn me about this? :-)

### PR DESCRIPTION
**Pull request by evanrittenhouse**
## Summary

Adds `RUF017`. Closes #5073 

## Test Plan

`cargo t`

### CODE REVIEW

File: crates/ruff/src/rules/ruff/rules/quadratic_list_summation.rs [hunk 1]
```diff
@@ -0,0 +1,114 @@
...
+        return;
+    }
+
+    let Arguments { args, .. } = arguments;
+    let Some(list_string_repr) = args
+        .first()
+        .and_then(|arg| match arg {
+            Expr::Name(ast::ExprName { id, .. }) => Some(id.as_str()),
+            Expr::List(ast::ExprList { range, .. }) => Some(checker.locator().slice(*range)),
+            _ => None,
```
Comment (konstin) [inline crates/ruff/src/rules/ruff/rules/quadratic_list_summation.rs:62]: i would merge this match with the match from `verify_start_arg`, either by having `verify_start_arg` return this value or by inlining the function

File: crates/ruff/src/rules/ruff/rules/quadratic_list_summation.rs [hunk 2]
```diff
@@ -0,0 +1,114 @@
...
+        return;
+    }
+
+    if verify_start_arg(arguments, checker).is_none() {
+        return;
+    }
+
+    let Arguments { args, .. } = arguments;
+    let Some(list_string_repr) = args
+        .first()
```
Comment (konstin) [inline crates/ruff/src/rules/ruff/rules/quadratic_list_summation.rs:58]: nit: use the following pattern for slightly more straightforward code
```suggestion
    let Some(first) = args.first() else {
        return;
    }

    match first {
        Expr::Name(ast::ExprName { id, .. }) => ...,
        Expr::List(ast::ExprList { range, .. }) => ...,
        _ => return,
    }
```

File: crates/ruff/resources/test/fixtures/ruff/RUF017.py [hunk 3]
```diff
@@ -0,0 +1,12 @@
+x = [1, 2, 3]
+y = [4, 5, 6]
+
+# RUF017
+sum([x, y], start=[])
+sum([x, y], [])
+sum([[1, 2, 3], [4, 5, 6]], start=[])
+sum([[1, 2, 3], [4, 5, 6]], [])
```
Comment (konstin) [inline crates/ruff/resources/test/fixtures/ruff/RUF017.py:8]: can you a case where the expression is multiline? your code should work but the case should be in the tests

Comment (evanrittenhouse) [reply crates/ruff/resources/test/fixtures/ruff/RUF017.py:8]: Added

Comment (konstin) [reply crates/ruff/resources/test/fixtures/ruff/RUF017.py:8]: Is it intentional that you didn't push the change yet?

Comment (evanrittenhouse) [reply crates/ruff/resources/test/fixtures/ruff/RUF017.py:8]: Should be right below this line

**Comment by charliermarsh** [comment]
Thanks! I tweaked this to use the `functools.reduce` approach suggested in the issue, since we _can_ import symbols and I find it to be more readable (closer to `sum()`). I also put this in the nursery for now, I'd like to get feedback on it before we upgrade it to production.

**Comment by charliermarsh** [comment]
Oh, I'm not able to push my edits. Do you mind enabling edits, @evanrittenhouse?

**Comment by evanrittenhouse** [comment]
Sounds good - enabled edits. Didn't know we are down for importing other modules - assuming that those are built-ins only, right? Are there any contexts where we wouldn't want to do that?

**Comment by charliermarsh** [comment]
@evanrittenhouse - Good question. We can import anything from the standard library (`functools` and `operator` are both standard-library modules), but we'd never want to import anything outside of the standard library.

File: crates/ruff/src/rules/ruff/snapshots/ruff__rules__ruff__tests__RUF017_RUF017.py.snap [hunk 4]
```diff
@@ -0,0 +1,53 @@
...
+7 9 | sum([[1, 2, 3], [4, 5, 6]], start=[])
+8 10 | sum([[1, 2, 3], [4, 5, 6]], [])
+
+RUF017.py:7:1: RUF017 [*] Avoid quadratic list summation
+  |
+5 | sum([x, y], start=[])
+6 | sum([x, y], [])
+7 | sum([[1, 2, 3], [4, 5, 6]], start=[])
+  | ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ RUF017
+8 | sum([[1, 2, 3], [4, 5, 6]], [])
```
Comment (hauntsaninja) [inline crates/ruff/src/rules/ruff/snapshots/ruff__rules__ruff__tests__RUF017_RUF017.py.snap:33]: Is it intentional this only complains when start is passed by kwarg?

Comment (charliermarsh) [reply crates/ruff/src/rules/ruff/snapshots/ruff__rules__ruff__tests__RUF017_RUF017.py.snap:33]: Just my mistake, @evanrittenhouse had it right initially. I looked at the docs and saw the `/` for the positional-only `iterable` argument, and my mind immediately went to keyword-only, wasn't thinking straight. PR welcome, or I'll fix a little later.

Comment (hauntsaninja) [reply crates/ruff/src/rules/ruff/snapshots/ruff__rules__ruff__tests__RUF017_RUF017.py.snap:33]: Thanks both! Easy fix, will open a PR

Comment (hauntsaninja) [reply crates/ruff/src/rules/ruff/snapshots/ruff__rules__ruff__tests__RUF017_RUF017.py.snap:33]: https://github.com/astral-sh/ruff/pull/6664

**Pull request merged**